src/hip/models.pyi

```diff
@@ -0,0 +1,81 @@
+import typing
+
+HeadersType = typing.Union[
+    typing.Mapping[str, str],
+    typing.Mapping[bytes, bytes],
+    typing.Iterable[typing.Tuple[str, str]],
+    typing.Iterable[typing.Tuple[bytes, bytes]],
+]
+
+class SyncRequestData:
+    """Represents a synchronous request data object. Basically a wrapper around whatever
+    a user passes in via 'data', 'files', 'json', etc parameters. We can take bytes /
+    strings, iterators of bytes or strings, and files. We can also sub-class / implement
+    a file-like interface for things like multipart-formdata so it's possible to
+    rewind (unlike the current urllib3 interface).
+
+    Maybe in the future we can expose details about when an object can be sent via
+    'socket.sendfile()', etc. This would have to be communicated somehow to the
+    low-level backend streams.
+
+    When we're handed a file-like object we should take down the starting point
+    in the file via .tell() so we can rewind and put the pointer back after
+    the seek() call in '.content_length'
+
+    Ref: https://github.com/python-trio/urllib3/issues/135
+    """
+
+    def read(self, nbytes: int) -> bytes:
```

Comment:
Regarding there only being one interface, I think we should have something like `SyncRequestData.sync_read()` and `AsyncRequestData.async_read()` or maybe we switch to using `async for chunk in request.data` which would automatically get transformed by unasync but we'd forfeit the ability to grab a certain amount of data. I guess that interface is easier to implement anyways?

The reason I think this would be good to have the split is you'd only have to subclass `AsyncRequestData` and could technically use that class in both async and synchronous requests. Would just need to have good error reporting if you're passed input data that is for example, an async generator, and then using it in a sync session. I'm guessing that wouldn't happen often and an error message would fix any ambiguity.

Reply:
Here's an example of my plan in action:

```python
import typing

StrOrInt = typing.Union[str, int]
FormType = typing.Union[
    typing.Sequence[typing.Tuple[str, StrOrInt]],
    typing.Mapping[str, typing.Union[StrOrInt, typing.Sequence[StrOrInt]]],
]


def _int_to_urlenc():
    values = {}
    special = {0x2A, 0x2D, 0x2E, 0x5F}
    for byte in range(256):
        if (
            (0x61 <= byte <= 0x7A)
            or (0x030 <= byte <= 0x5A and byte != 0x40)
            or (byte in special)
        ):  # Keep the ASCII
            values[byte] = bytes((byte,))
        elif byte == 0x020:  # Space -> '+'
            values[byte] = b"+"
        else:  # Percent-encoded
            values[byte] = b"%" + hex(byte)[2:].upper().encode()
    return values


INT_TO_URLENC = _int_to_urlenc()


class URLEncoded(AsyncRequestData):
    """Implements x-www-form-urlencoded as a RequestData object"""
    def __init__(self, form: FormType):
        self._form = form
        self._data = None

    @property
    def content_type(self) -> str:
        return "application/x-www-form-urlencoded"

    @property
    def content_length(self) -> int:
        return len(self._encode_form())

    def is_rewindable(self) -> bool:
        return True

    def __iter__(self) -> typing.Iterable[bytes]:
        yield self._encode_form()

    async def __aiter__(self) -> typing.AsyncIterable[bytes]:
        yield self._encode_form()

    def _encode_form(self) -> bytes:
        if self._data is None:

            def serialize(x: str) -> bytes:
                return b"".join([INT_TO_URLENC[byte] for byte in x.encode("utf-8")])

            output: typing.List[bytes] = []
            for k, vs in (
                self._form.items() if hasattr(self._form, "items") else self._form
            ):
                if isinstance(vs, str) or not hasattr(vs, "__iter__"):
                    vs = (vs,)
                for v in vs:
                    output.append(serialize(k) + b"=" + serialize(v))

            self._data = b"&".join(output)

        return self._data

```

Suddenly this one class can be used for both synchronous and asynchronous requests. We keep the `is_rewindable` (or maybe rename it to `is_replayable`? or something else) and then we'd signal a "rewind" by calling `__aiter__` or `__iter__` again.